chia/full_node/full_node.py

```diff
@@ -1771,7 +1771,7 @@ async def respond_compact_proof_of_time(self, request: timelord_protocol.Respond
             request.vdf_info, request.vdf_proof, request.height, request.header_hash, field_vdf
         ):
             return None
-        async with self.blockchain.lock:
+        async with self.blockchain.compact_proof_lock:
             await self._replace_proof(request.vdf_info, request.vdf_proof, request.height, field_vdf)
```

Comment:
how come we don't check the `seen_compact_proof` here?

Reply:
This is a message directly from the timelord so there is not much reason to check this, since we only have 1 timelord connection most of the time

Reply:
the function, `_replace_proof()` calls:
```
self.block_store.get_full_blocks_at([height])
```

The surely is sensitive to reorgs. that might return different blocks for the same height. We then call:

```
self.blockchain.get_block_record_from_db()
```

When we're done, we write back to the database with:
```
self.block_store.add_full_block(new_block, block_record)
self.block_store.db_wrapper.commit_transaction()
```

Are all of those really safe to do without holding the blockchain lock?

Also, it doesn't look like we *start* the transaction, we just commit it at the end..

Reply:
Good point, if these compact blocks are at a very old height, it's fine, but if they are recent, there might be a reorg happning (in which case we would need the blockchain lock.)

This is fine, a new transaction will get started automatically when writing if there is no transaction

Reply:
Oh wait no, it's not an issue. We are holding the db_wrapper lock already.

Reply:
as far as I can tell, the db wrapper lock doesn't really do anything. We don't hold it for the complete operation, just for individual db calls

Reply:
we replace the proofs for all blocks of a certain height and we don't delete blocks from the database.
so it seems to me that the worst case here if this happens during a reorg is that some of the blocks are not in the database yet and will not have the proof replaced.

i think the issue here is that we replace for all blocks of a certain height but we only check that the compact proof matches the block in our current chain

Reply:
@fchirica  can we  just call _can_accept_compact_proof on each block before we replace proof ?

Reply:
and blocks are immutable, once in the DB, so even if the block whose proof we're replacing gets reorged out of the current chain, that doesn't matter.

I suppose blocks are immutable except for the proofs can be compacted. The new lock you introduced makes sure we never replace the proof of the same block at the same time, right?

Reply:
we should not do concurrent replacing at all

Reply:
sorry, we do check the proof match but only for the end of slots

Reply:
We finally understood what this code does. It replaces one VDF at that height, and checks all blocks present at that height. When if finds the one that matches, it replaces it. It is checking for matches for 2/4 of the VDF types, so this is a bug. It needs to check it for all types:
https://github.com/Chia-Network/chia-blockchain/blob/main/chia/full_node/full_node.py#L1764

Reply:
yes, added the checks for ip and sp